sky/execution.py

```diff
@@ -451,15 +456,47 @@ def launch(
         controller_utils.check_cluster_name_not_controller(
             cluster_name, operation_str='sky.launch')
 
+    handle = None
+    stages = None
+    # Check if cluster exists and we are doing fast provisioning
+    if fast and cluster_name is not None:
+        maybe_handle = global_user_state.get_handle_from_cluster_name(
+            cluster_name)
+        if maybe_handle is not None:
+            try:
+                # This will throw if the cluster is not available
+                backend_utils.check_cluster_available(
+                    cluster_name,
+                    operation='executing tasks',
+                    check_cloud_vm_ray_backend=False,
+                    dryrun=dryrun)
+                # If the cluster is available, restrict stages
+                handle = maybe_handle
+                stages = [
+                    # Stage.CLONE_DISK,
+                    # Stage.PROVISION,
+                    # Stage.OPTIMIZE,
+                    Stage.SYNC_WORKDIR,
+                    # Stage.SYNC_FILE_MOUNTS,
+                    # Stage.SETUP,
+                    # Stage.PRE_EXEC,
+                    Stage.EXEC,
+                    # Stage.DOWN
+                ]
```

Comment:
For readability and maintainability, should we directly invoke and return `exec()` after L468?

We may need to add a `_disable_controller_check` arg to skip the controller_utils.check_cluster_name_not_controller() check:
https://github.com/skypilot-org/skypilot/blob/c10fd4262c36d5a008b5aaa6c79d6efe3fad5c0c/sky/execution.py#L544-L545

Reply:
I think we don't really want to go that direction - although that is the simple way of thinking about it for now, this logic will become more complicated and diverge from exec over time.